python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -271,28 +313,222 @@ def latest_materialization_records_by_key(
+        from_partitions: Optional[PartitionsDefinition] = None,
+    ) -> Sequence[str]:
+        """
+        Given two partitions definitions, converts a partition key in from_partitions to the
+        corresponding partition in to_partitions. Matches based on the time window the partition
+        belongs in.
+        """
+
+        partition_key = check.str_param(partition_key, "partition_key")
+        to_partitions = check.inst_param(to_partitions, "to_partitions", PartitionsDefinition)
+        from_partitions = check.opt_inst_param(
+            from_partitions, "from_partitions", PartitionsDefinition
+        )
+
+        if from_partitions == None and len(self._asset_keys) == 1:
+            from_partitions = self._partitions_def_by_asset_key.get(self._asset_keys[0])
+        elif not self._asset_keys or len(self._asset_keys) != 1:
+            raise DagsterInvalidInvocationError(
+                "from_partitions is a required parameter unless if there is only one asset provided"
+                "to the sensor context"
+            )
+
+        if not isinstance(from_partitions, PartitionsDefinition):
+            raise DagsterInvalidInvocationError(
+                "from_partitions must be provided if assets are not provided or are unpartitioned"
+            )
+        else:
+            if not self._can_map(to_partitions, from_partitions):
+                raise DagsterInvalidInvocationError(
+                    "Currently only time window partitions are supported"
+                )
+
+            # Currently, only time window partitions are supported
+            # TODO: support providing partition mappings for other partition types
+            downstream_partition_key_range = (
+                TimeWindowPartitionMapping().get_downstream_partitions_for_partition_range(
+                    PartitionKeyRange(partition_key, partition_key),
+                    downstream_partitions_def=to_partitions,
+                    upstream_partitions_def=from_partitions,
+                )
+            )
+            partition_keys = to_partitions.get_partition_keys()
+            if (
+                downstream_partition_key_range.start not in partition_keys
+                or downstream_partition_key_range.end not in partition_keys
+            ):
+                raise DagsterInvalidInvocationError(
+                    f"Mapped partition key {partition_key} to downstream partition key range "
+                    f"[{downstream_partition_key_range.start}...{downstream_partition_key_range.end}] "
+                    "which is not a valid range in the downstream partitions definition."
+                )
+
+            downstream_partitions = partition_keys[
+                partition_keys.index(downstream_partition_key_range.start) : partition_keys.index(
+                    downstream_partition_key_range.end
+                )
+                + 1
+            ]
+            return downstream_partitions
+
+    @public
```

Comment:
Does this need to be public?

Reply:
I guess what I mean is: does it need to be user-facing"

Reply:
I think it would be a helpful utility method if users want to fetch the partition off an event log record

Reply:
Got it. The reason I'm pushing back against this is that it feels awkward to me for this to be a method of the sensor context, given that it doesn't use `self` at all.  If anything, it seems like maybe it should be a method of the EventLogRecord?  Although, by that logic, we might end up adding a lot of other methods to EventLogRecord to get other assorted data within the event? Or we could provide a method that returns a mapping of (asset key, partition key) to EventLogRecord?  Or, if we don't think that users will need this functionality in common cases, we could keep it private to punt on figuring it out.

Reply:
I think we should add these methods to `EventLogRecord` eventually. But in terms of the multi asset sensor use case, I think we can just make this method private. In most cases I think users will use `latest_materialization_records_by_partition`, which will return a dict of partition key -> materialization, so we don't need to make this utility available